File: day56/Trie.py
class Trie:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.root = {}
        self.end_of_word = "#"

    def insert(self, word: str) -> None:
        """
        Inserts a word into the trie.
        """
        node = self.root
        for char in word:
            node = node.setdefault(char, {})
        node[self.end_of_word] = self.end_of_word

    def search(self, word: str) -> bool:
        """
        Returns if the word is in the trie.
        """
        node = self.root
        for char in word:
            if char not in node:
                return False
            node = node[char]
        return self.end_of_word in node

    def startsWith(self, prefix: str) -> bool:
        """
        Returns if there is any word in the trie that starts with the given prefix.
        """
        node = self.root
        for char in prefix:
            if char not in node:
                return False
            node = node[char]
        return True

# class Trie:

#     def __init__(self):
#         """
#         Initialize your data structure here.
#         """
#         self.sub_node = [None] * 26
#         self.isEnd = False

#     def insert(self, word: str) -> None:
#         """
#         Inserts a word into the trie.
#         """
#         node = self
#         for char in word:
#             index = ord(char) - ord("a")
#             if node.sub_node[index] == None: node.sub_node[index] = Trie();
#             node = node.sub_node[index]
#         node.isEnd = True

#     def search(self, word: str) -> bool:
#         """
#         Returns if the word is in the trie.
#         """
#         node = self
#         for char in word:
#             index = ord(char) - ord("a")
#             if node.sub_node[index] == None: return False
#             node = node.sub_node[index]
#         return node.isEnd

#     def startsWith(self, prefix: str) -> bool:
#         """
#         Returns if there is any word in the trie that starts with the given prefix.
#         """
#         node = self
#         for char in prefix:
#             index = ord(char) - ord("a")
#             if node.sub_node[index] == None: return False
#             node = node.sub_node[index]
#         return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: day56/Trie.java
class Trie {

    private class TrieNode {
        private TrieNode[] next = new TrieNode[26];
        private boolean isEnd = false;

        public TrieNode() { }
    }

    private TrieNode root;

    /** Initialize your data structure here. */
    public Trie() {
        this.root = new TrieNode();
    }
    
    /** Inserts a word into the trie. */
    public void insert(String word) {
        TrieNode cur = this.root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (cur.next[index] == null) cur.next[index] = new TrieNode();
            cur = cur.next[index];
        }
        cur.isEnd = true;

    }
    
    /** Returns if the word is in the trie. */
    public boolean search(String word) {
        TrieNode cur = this.root;
        for (int i = 0; i < word.length(); i++) {
            int index = word.charAt(i) - 'a';
            if (cur.next[index] == null) return false;
            cur = cur.next[index];
        }
        return cur.isEnd;

    }
    
    /** Returns if there is any word in the trie that starts with the given prefix. */
    public boolean startsWith(String prefix) {
        TrieNode cur = this.root;
        for (int i = 0; i < prefix.length(); i++) {
            int index = prefix.charAt(i) - 'a';
            if (cur.next[index] == null) return false;
            cur = cur.next[index];
        }
        return true;

    }
}

// class Trie {
//     private Trie[] next = new Trie[26];
//     private boolean isEnd = false;

//     /** Initialize your data structure here. */
//     public Trie() {
        
//     }
    
//     /** Inserts a word into the trie. */
//     public void insert(String word) {
//         Trie root = this;
//         for (int i = 0; i < word.length(); i++) {
//             int index = word.charAt(i) - 'a';
//             if (root.next[index] == null) root.next[index] = new Trie();
//             root = root.next[index];
//         }
//         root.isEnd = true;

//     }
    
//     /** Returns if the word is in the trie. */
//     public boolean search(String word) {
//         Trie root = this;
//         for (int i = 0; i < word.length(); i++) {
//             int index = word.charAt(i) - 'a';
//             if (root.next[index] == null) return false;
//             root = root.next[index];
//         }
//         return root.isEnd;

//     }
    
//     /** Returns if there is any word in the trie that starts with the given prefix. */
//     public boolean startsWith(String prefix) {
//         Trie root = this;
//         for (int i = 0; i < prefix.length(); i++) {
//             int index = prefix.charAt(i) - 'a';
//             if (root.next[index] == null) return false;
//             root = root.next[index];
//         }
//         return true;

//     }
// }

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */
